src/validation.cpp

```diff
@@ -2295,13 +2304,15 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
 
             // TODO: double parsing of the transaction. Should be done more efficiently.
             if (fStakingActive) {
+                CStakingTransactionParser stakingTxParser(MakeTransactionRef(tx));
                 if (txfee == 0) {
-                    if (!CheckIfEligibleFreeTx(tx, view, stakes, pindex->nHeight, chainparams.GetConsensus())) {
+                    if (!CheckIfEligibleFreeTx(tx, view, stakes, pindex->nHeight, chainparams.GetConsensus(),
+                            fStakingWithdrawal || stakingTxParser.GetStakingTxType() == StakingTransactionType::BURN ||
+                            stakingTxParser.GetStakingTxType() == StakingTransactionType::DEPOSIT)) {
```

Comment:
TBH That quite ugly if

Reply:
Maybe more brackets with braking it over few lines would help to make it more readable